app/src/main/java/nodomain/freeyourgadget/gadgetbridge/activities/HeartRateUtils.java

```diff
@@ -28,7 +32,42 @@
      */
     public static final int MAX_HR_MEASUREMENTS_GAP_MINUTES = 10;
 
-    public static boolean isValidHeartRateValue(int value) {
-        return value > HeartRateUtils.MIN_HEART_RATE_VALUE && value < HeartRateUtils.MAX_HEART_RATE_VALUE;
+    private int maxHeartRateValue;
+    private int minHeartRateValue;
+
+    /**
+     * extra class used for synchronized access of the singleton
+     */
+    private static final class HeartRateUtilsInstanceHolder {
```

Comment:
Can you please elaborate why we need the separate class?

If you put the static initialization directly into the `HeartRateUtils` class, this will happen at classloading time, which is perfectly synchronized. Could the static method be invoked before the static initializer is finished?

Reply:
This is used to be synchronized when calling HeartRateUtils.getInstance() the first time in two different threads (race condition).
This will lazy load the instance when the class is loaded.

Reply:
AFAICS we don't need any synchronization here. If we would lazily create the instance inside the `getInstance()` method, then this would need to be synchronized. But we create the instance at classloading time already (in the static initializer), so no need for any further synchronization, nor for the inner class.